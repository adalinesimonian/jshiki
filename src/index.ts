import Evaluator, {
  UnaryOperatorOptions,
  BinaryOperatorOptions,
  LogicalOperatorOptions,
  OperatorOptions,
  EvaluatorOptions,
} from './evaluator'
import getRuleProxy from './rule-proxy'
import RuleTree, {
  AccessPath,
  AccessRule,
  AllowAccessRule,
  BlockAccessRule,
} from './rule-tree'

export {
  AccessPath,
  AccessRule,
  AllowAccessRule,
  BlockAccessRule,
  UnaryOperatorOptions,
  BinaryOperatorOptions,
  LogicalOperatorOptions,
  OperatorOptions,
}

/**
 * Options for parsing.
 */
export interface JshikiParseOptions extends EvaluatorOptions {
  /**
   * Access rules to use when determining what properties can be accessed by the
   * expression. The rules are evaluated in order.
   * @example
   * ```js
   * const options = {
   *   rules: [
   *     { allow: 'foo.bar' },
   *     { block: ['baz', 'qux'] },
   *   ]
   * }
   * const obj = {
   *   foo: { bar: 'baz' },
   *   baz: { qux: 'quux' },
   * }
   * let expr = jshiki.parse('foo.bar', options)
   * expr(obj) // => 'baz'
   *
   * expr = jshiki.parse('baz.qux', options)
   * expr(obj) // => undefined
   * ```
   */
  rules?: AccessRule[]
  /**
   * If true, only properties with a matching allow rule can be accessed by the
   * expression. If false, all properties can be accessed unless they have a
   * block rule. Defaults to false.
   * @default false
   * @example
   * ```js
   * const options = {
   *   explicitAllow: true,
   *   rules: [{ allow: 'foo.bar' }]
   * }
   * const obj = {
   *   foo: { bar: 'baz' },
   *   baz: { qux: 'quux' },
   * }
   * let expr = jshiki.parse('foo.bar', options)
   * expr(obj) // => 'baz'
   *
   * let expr = jshiki.parse('baz', options)
   * expr(obj) // => undefined
   * ```
   */
  explicitAllow?: boolean
}

/**
 * Options for evaluating.
 */
export interface JshikiEvaluateOptions extends JshikiParseOptions {
  /**
   * The scope to use when evaluating the expression. The expression will be
   * limited to accessing the properties of the scope.
   * @example
   * ```js
   * const options = {
   *   scope: {
   *     foo: 'bar',
   *     baz: {
   *       qux: 'quux',
   *     }
   *   }
   * }
   * jshiki.evaluate('foo', options) // => 'bar'
   * jshiki.evaluate('baz.qux', options) // => 'quux'
   * ```
   */
  scope?: Record<any, any>
}

/**
 * A compiled, executable expression.
 *
 * ```js
 * const expression = jshiki.parse('foo.bar')
 * expression({
 *   foo: { bar: 'baz' },
 * }) // => 'baz'
 * ```
 * @param scope The scope to use when evaluating the expression. The expression
 * will be limited to accessing the properties of the scope.
 */
export type JshikiExpression = (scope?: Record<any, any>) => any

/**
 * Parses an expression into an executable function.
 * @param str The expression to parse.
 * @param options The options to use.
 * @returns The parsed expression.
 * @example
 * ```js
 * const expression = jshiki.parse('(a || 1) + 2')
 * let result = expression()
 * // result => 3
 * result = expression({ a: 5 })
 * // result => 7
 * ```
 */
export function parse(
  str: string,
  { rules, explicitAllow, operators, expressions }: JshikiParseOptions = {}
): JshikiExpression {
  const ruleTree = rules ? new RuleTree(rules) : undefined
  const expression = new Evaluator({ operators, expressions }).createExpression(
    str
  )

  return (scope: Record<any, any> = {}) => {
    const proxiedScope = ruleTree
      ? getRuleProxy(scope, ruleTree, Boolean(explicitAllow))
      : scope
    return expression(proxiedScope)
  }
}

/**
 * Evaluates an expression.
 * @param str The expression to evaluate.
 * @param options The options to use.
 * @returns The result of the expression.
 * @example
 * ```js
 * let result = jshiki.evaluate('1 + 2')
 * // result => 3
 * result = jshiki.evaluate('a + 2', { scope: { a: 5 } })
 * // result => 7
 * ```
 */
export function evaluate(str: string, options?: JshikiEvaluateOptions): any {
  return parse(str, options)(options?.scope)
}

export default { parse, evaluate }
