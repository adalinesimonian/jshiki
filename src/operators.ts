export const unary: { [operator: string]: (o: any) => any } = {
  '+': (o: any): number => +o,
  '-': (o: any): number => -o,
  '!': (o: any): boolean => !o,
  '~': (o: any): number => ~o,
  typeof: (o: any): string => typeof o,
}

export const binary: { [operator: string]: (l: any, r: any) => any } = {
  '+': (l: any, r: any): any => l + r,
  '-': (l: any, r: any): number => l - r,
  '*': (l: any, r: any): number => l * r,
  '**': (l: any, r: any): number => l ** r,
  '/': (l: any, r: any): number => l / r,
  '%': (l: any, r: any): number => l % r,
  '<': (l: any, r: any): boolean => l < r,
  '>': (l: any, r: any): boolean => l > r,
  '<=': (l: any, r: any): boolean => l <= r,
  '>=': (l: any, r: any): boolean => l >= r,
  /* eslint-disable eqeqeq */
  '==': (l: any, r: any): boolean => l == r,
  '!=': (l: any, r: any): boolean => l != r,
  /* eslint-enable eqeqeq */
  '===': (l: any, r: any): boolean => l === r,
  '!==': (l: any, r: any): boolean => l !== r,
  '|': (l: any, r: any): number => l | r,
  '^': (l: any, r: any): number => l ^ r,
  '&': (l: any, r: any): number => l & r,
  '<<': (l: any, r: any): number => l << r,
  '>>': (l: any, r: any): number => l >> r,
  '>>>': (l: any, r: any): number => l >>> r,
  in: (l: any, r: any): boolean => l in r,
  instanceof: (l: any, r: any): boolean => l instanceof r,
}

export const logical: {
  [operator: string]: {
    (l: () => any, r: () => any): any
    async: (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => Promise<{ result: any }>
  }
} = {
  '||': Object.assign((l: () => any, r: () => any): any => l() || r(), {
    async: async (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => ({ result: (await l()).value || (await r()).value }),
  }),
  '&&': Object.assign((l: () => any, r: () => any): any => l() && r(), {
    async: async (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => ({ result: (await l()).value && (await r()).value }),
  }),
  '??': Object.assign((l: () => any, r: () => any): any => l() ?? r(), {
    async: async (
      l: () => Promise<{ value: any }>,
      r: () => Promise<{ value: any }>
    ) => ({ result: (await l()).value ?? (await r()).value }),
  }),
}

export default {
  unary,
  binary,
  logical,
}
