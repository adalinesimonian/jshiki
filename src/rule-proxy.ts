import RuleTree, { AccessPathArray, AccessPathPart } from './rule-tree'

function hasAccess(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  propertyScope: AccessPathArray
) {
  const match = ruleTree.match(propertyScope)
  return explicitAllow
    ? match === true || (typeof match === 'string' && match === 'allow')
    : !(typeof match === 'string' && match === 'block')
}

function getProxyHandler<T extends object>(
  ruleTree: RuleTree,
  explicitAllow: boolean,
  scope: AccessPathArray = []
): ProxyHandler<T> {
  const cache = new Map<AccessPathPart, any>()

  return {
    get(target: T, property: AccessPathPart) {
      if (cache.has(property)) {
        return cache.get(property)
      }

      const propertyScope = scope.concat(property)

      if (!hasAccess(ruleTree, explicitAllow, propertyScope)) {
        cache.set(property, undefined)
        return undefined
      }

      const propValue = (target as any)[property]
      const returnValue =
        (propValue && typeof propValue === 'object') ||
        typeof propValue === 'function'
          ? new Proxy(
              (target as any)[property],
              getProxyHandler(ruleTree, explicitAllow, propertyScope)
            )
          : propValue
      cache.set(property, returnValue)
      return returnValue
    },

    set(_target: T, _property: AccessPathPart, value: any) {
      return value
    },
  }
}

export default function getRuleProxy<T extends object>(
  target: T,
  ruleTree: RuleTree,
  explicitAllow?: boolean
): T {
  return (target && typeof target === 'object') || typeof target === 'function'
    ? new Proxy(target, getProxyHandler(ruleTree, Boolean(explicitAllow)))
    : target
}
