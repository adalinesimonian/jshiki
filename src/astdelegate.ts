import {
  ArrayExpression,
  BinaryExpression,
  CallExpression,
  ConditionalExpression,
  Delegate,
  Expression,
  IdentifierExpression,
  LiteralExpression,
  MemberExpression,
  ObjectExpression,
  PropertyExpression,
  Syntax,
  Token,
  TopLevelExpression,
  UnaryExpression,
} from './lib/esprima/types'
import operators from './operators'

export interface ASTDelegateOptions {
  /**
   * The scope to use when evaluating the expression. The expression will be
   * limited to accessing the properties of the scope.
   */
  scope?: Record<any, any>
}

export default class ASTDelegate implements Delegate {
  scope: any

  constructor({ scope }: ASTDelegateOptions = {}) {
    this.scope = scope || {}
  }

  #evaluateIdentifier(expression: IdentifierExpression): any {
    const { identifier } = expression
    if (typeof identifier === 'string') {
      return this.scope[identifier]
    }
    return identifier
  }

  createUnaryExpression(op: string, arg: Expression): UnaryExpression {
    if (!operators.unary[op as keyof typeof operators.unary]) {
      throw Error(`Disallowed operator: ${op}`)
    }

    return Object.assign(
      () => operators.unary[op as keyof typeof operators.unary](arg()),
      {
        type: Syntax.UnaryExpression as const,
        operator: op,
        argument: arg,
      }
    )
  }

  createBinaryExpression(
    op: string,
    left: Expression,
    right: Expression
  ): BinaryExpression {
    if (!operators.binary[op as keyof typeof operators.binary]) {
      throw Error(`Disallowed operator: ${op}`)
    }

    const leftExpr =
      left.type === Syntax.Identifier
        ? () => this.#evaluateIdentifier(left)
        : left
    const rightExpr =
      right.type === Syntax.Identifier
        ? () => this.#evaluateIdentifier(right)
        : right

    return Object.assign(
      () =>
        operators.binary[op as keyof typeof operators.binary](
          leftExpr(),
          rightExpr()
        ),
      {
        type: Syntax.BinaryExpression as const,
        operator: op,
        left,
        right,
      }
    )
  }

  createConditionalExpression(
    test: Expression,
    consequent: Expression,
    alternate: Expression
  ): ConditionalExpression {
    return Object.assign(() => (test() ? consequent() : alternate()), {
      type: Syntax.ConditionalExpression as const,
      test,
      consequent,
      alternate,
    })
  }

  createIdentifier(identifier: string | boolean | null): IdentifierExpression {
    return Object.assign(() => identifier, {
      type: Syntax.Identifier as const,
      identifier: identifier,
    })
  }

  createMemberExpression(
    accessor: '.' | '[',
    object: Expression,
    property: IdentifierExpression
  ): MemberExpression {
    const objectEvaluator =
      object.type === Syntax.Identifier
        ? () => this.#evaluateIdentifier(object)
        : object

    return Object.assign(() => objectEvaluator()[property()], {
      type: Syntax.MemberExpression as const,
      accessor,
      object,
      property,
    })
  }

  createCallExpression(callee: Expression, args: Expression[]): CallExpression {
    return Object.assign(
      () => {
        let func, self
        if (callee.type === Syntax.Identifier) {
          self = this.#evaluateIdentifier(callee)
          func = self
        } else if (callee.type === Syntax.MemberExpression) {
          self = callee.object()
          func = callee()
        } else {
          self = callee()
          func = callee
        }
        return func.apply(
          self,
          args.map(arg => arg())
        )
      },
      {
        type: Syntax.CallExpression as const,
        callee,
        args,
      }
    )
  }

  createLiteral(
    token: Pick<
      Extract<Token, { value: string | number | boolean | null }>,
      'value'
    >
  ): LiteralExpression {
    return Object.assign(() => token.value, {
      type: Syntax.Literal as const,
      value: token.value,
    })
  }

  createArrayExpression(elements: (Expression | null)[]): ArrayExpression {
    return Object.assign(() => elements.map(element => element && element()), {
      type: Syntax.ArrayExpression as const,
      elements,
    })
  }

  createProperty(
    kind: 'init',
    key: IdentifierExpression,
    value: Expression
  ): PropertyExpression {
    return Object.assign(
      () => ({
        key: key({ child: true }),
        value: value(),
      }),
      {
        type: Syntax.Property as const,
        kind,
        key,
        value,
      }
    )
  }

  createObjectExpression(properties: Expression[]): ObjectExpression {
    return Object.assign(
      () => {
        const object: Record<any, any> = {}
        for (const property of properties) {
          const { key, value } = property()
          object[key] = value
        }
        return object
      },
      {
        type: Syntax.ObjectExpression as const,
        properties,
      }
    )
  }

  createTopLevel(expression: Expression): TopLevelExpression {
    let evaluator
    if (expression.type === Syntax.Identifier) {
      const identifier = this.#evaluateIdentifier(expression)
      evaluator = () => identifier
    } else {
      evaluator = expression
    }

    return Object.assign(evaluator, {
      type: Syntax.TopLevel as const,
      expression,
    })
  }

  createThisExpression(): never {
    // TODO
    throw new Error('`this` is not supported')
  }
}
