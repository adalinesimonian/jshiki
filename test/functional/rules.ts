import { evaluate, evaluateAsync } from '../../src'

const symbolBeta = Symbol('beta')
function getTestObject() {
  return {
    symbolBeta,
    x: {
      y: {
        z: 'foo',
      },
    },
    a: {
      b: {
        c: 'bar',
      },
    },
    alpha: {
      [symbolBeta]: 'baz',
      beta: {
        gamma: 'qux',
      },
    },
    foo: {
      bar: {
        baz: 'quux',
      },
    },
    wilds: {
      '*': 'quuz',
      '**': 'corge',
    },
    shallow: {
      deep: {
        deeper: {
          deepest: {
            grault: 'garply',
            waldo: 'fred',
          },
        },
      },
    },
  }
}

describe('Rules', () => {
  describe('with explicit allow set to true', () => {
    it('should not be able to access properties w/o matching rule', () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }, { allow: 'symbolBeta' }],
        scope: getTestObject(),
      }
      expect(evaluate('foo', options)).toBeUndefined()
      expect(() => evaluate('foo.bar', options)).toThrowError()
      expect(() => evaluate('foo["b" + "a" + "r"]', options)).toThrowError()
      expect(() => evaluate('alpha[symbolBeta]', options)).toThrowError()
    })

    it('should be able to access properties w/ matching allow rule', () => {
      const options = {
        explicitAllow: true,
        rules: [
          { allow: 'x.y.z' },
          { block: 'a.b' },
          { allow: 'symbolBeta' },
          { allow: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(evaluate('x.y.z', options)).toBe('foo')
      expect(evaluate('x["y"]["z"]', options)).toBe('foo')
      expect(evaluate('alpha[symbolBeta]', options)).toBe('baz')
    })

    it('should not be able to access properties w/ matching block rule', () => {
      const options = {
        explicitAllow: true,
        rules: [
          { allow: 'a' },
          { block: 'a.b' },
          { allow: 'symbolBeta' },
          { allow: 'alpha' },
          { block: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(evaluate('a.b', options)).toBeUndefined()
      expect(evaluate('a["b"]', options)).toBeUndefined()
      expect(() => evaluate('a.b.c', options)).toThrowError()
      expect(() => evaluate('a["b"]["c"]', options)).toThrowError()
      expect(evaluate('alpha[symbolBeta]', options)).toBeUndefined()
    })

    it('should return objects that return the same instances on subsequent property access', () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = evaluate('x', options)
      const y = result.y
      expect(y).toBe(result.y)
    })

    it('should return objects with useless setters', () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = evaluate('x.y', options)
      result.z = 'qux'
      expect(evaluate('x.y.z', options)).toBe('foo')
    })

    it('should ignore rules on non-object, non-function scopes', () => {
      const options = {
        explicitAllow: true,
        rules: [],
        scope: 'str' as any,
      }
      expect(evaluate('length', options)).toBe(3)
    })
  })

  describe('with explicit allow set to false', () => {
    it('should be able to access properties w/o matching rule', () => {
      const options = {
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      expect(evaluate('foo', options)).toEqual({ bar: { baz: 'quux' } })
      expect(evaluate('foo.bar', options)).toEqual({ baz: 'quux' })
      expect(evaluate('foo["b" + "a" + "r"]', options)).toEqual({ baz: 'quux' })
      expect(evaluate('alpha[symbolBeta]', options)).toBe('baz')
    })

    it('should be able to access properties w/ matching allow rule', () => {
      const options = {
        rules: [
          { allow: 'x.y.z' },
          { block: 'a.b' },
          { allow: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(evaluate('x.y.z', options)).toBe('foo')
      expect(evaluate('x["y"]["z"]', options)).toBe('foo')
      expect(evaluate('alpha[symbolBeta]', options)).toBe('baz')
    })

    it('should not be able to access properties w/ matching block rule', () => {
      const options = {
        rules: [
          { allow: 'a' },
          { block: 'a.b' },
          { block: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(evaluate('a.b', options)).toBeUndefined()
      expect(evaluate('a["b"]', options)).toBeUndefined()
      expect(() => evaluate('a.b.c', options)).toThrowError()
      expect(() => evaluate('a["b"]["c"]', options)).toThrowError()
      expect(evaluate('alpha[symbolBeta]', options)).toBeUndefined()
    })

    it('should return objects that return the same instances on subsequent property access', () => {
      const options = {
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = evaluate('x', options)
      const y = result.y
      expect(y).toBe(result.y)
    })

    it('should return objects with useless setters', () => {
      const options = {
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = evaluate('x.y', options)
      result.z = 'qux'
      expect(evaluate('x.y.z', options)).toBe('foo')
    })

    it('should ignore rules on non-object, non-function scopes', () => {
      const options = {
        rules: [{ block: 'length' }],
        scope: 'str' as any,
      }
      expect(evaluate('length', options)).toBe(3)
    })
  })

  describe('Rule parsing', () => {
    it('should support escaped wildcards', () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'wilds.\\*' }, { allow: 'wilds.\\**' }],
        scope: getTestObject(),
      }
      expect(evaluate('wilds["*"]', options)).toBe('quuz')
      expect(evaluate('wilds["**"]', options)).toBe('corge')
    })

    it('should ignore falsy rules', () => {
      const options = {
        rules: [undefined as any],
        scope: getTestObject(),
      }
      expect(() => evaluate('x.y.z', options)).not.toThrow()
    })

    it('should ignore rules with empty paths', () => {
      const options = {
        rules: [{ block: '' }, { block: [] }],
        scope: getTestObject(),
      }
      expect(() => evaluate('x.y.z', options)).not.toThrow()
    })

    it('should respect rule order', () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'a.b.c' }, { block: 'a.b' }, { allow: 'a.b' }],
        scope: getTestObject(),
      }
      expect(evaluate('a.b.c', options)).toBeUndefined()
      expect(evaluate('a.b', options)).toEqual({ c: undefined })
    })

    it('should support wildcards', () => {
      const options = {
        explicitAllow: true,
        rules: [
          { allow: 'a.**' },
          { block: 'a.*.c' },
          { allow: 'shallow.**' },
          { block: 'shallow.**.grault' },
          { allow: 'shallow.**.deepest.waldo' },
        ],
        scope: getTestObject(),
      }
      expect(evaluate('a.b.c', options)).toBeUndefined()
      expect(evaluate('a.b', options)).toEqual({ c: undefined })
      expect(
        evaluate('shallow.deep.deeper.deepest.grault', options)
      ).toBeUndefined()
      expect(evaluate('shallow.deep.deeper.deepest.waldo', options)).toBe(
        'fred'
      )
    })

    it("should throw an error if a rule's path is not a string or string array", () => {
      const options = {
        rules: [{ allow: true as any }],
        scope: getTestObject(),
      }
      expect(() => evaluate('x.y.z', options)).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error if a rule's path part is not a string or a symbol", () => {
      const options = {
        rules: [{ allow: ['x', true as any] }],
        scope: getTestObject(),
      }
      expect(() => evaluate('x.y.z', options)).toThrowErrorMatchingSnapshot()
    })
  })
})

describe('Rules (async)', () => {
  describe('with explicit allow set to true', () => {
    it('should not be able to access properties w/o matching rule', async () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }, { allow: 'symbolBeta' }],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('foo', options)).toBeUndefined()
      expect(async () =>
        evaluateAsync('foo.bar', options)
      ).rejects.toThrowError()
      expect(async () =>
        evaluateAsync('foo["b" + "a" + "r"]', options)
      ).rejects.toThrowError()
      expect(async () =>
        evaluateAsync('alpha[symbolBeta]', options)
      ).rejects.toThrowError()
    })

    it('should be able to access properties w/ matching allow rule', async () => {
      const options = {
        explicitAllow: true,
        rules: [
          { allow: 'x.y.z' },
          { block: 'a.b' },
          { allow: 'symbolBeta' },
          { allow: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('x.y.z', options)).toBe('foo')
      expect(await evaluateAsync('x["y"]["z"]', options)).toBe('foo')
      expect(await evaluateAsync('alpha[symbolBeta]', options)).toBe('baz')
    })

    it('should not be able to access properties w/ matching block rule', async () => {
      const options = {
        explicitAllow: true,
        rules: [
          { allow: 'a' },
          { block: 'a.b' },
          { allow: 'symbolBeta' },
          { allow: 'alpha' },
          { block: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('a.b', options)).toBeUndefined()
      expect(await evaluateAsync('a["b"]', options)).toBeUndefined()
      expect(async () => evaluateAsync('a.b.c', options)).rejects.toThrowError()
      expect(async () =>
        evaluateAsync('a["b"]["c"]', options)
      ).rejects.toThrowError()
      expect(await evaluateAsync('alpha[symbolBeta]', options)).toBeUndefined()
    })

    it('should return objects that return the same instances on subsequent property access', async () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = await evaluateAsync('x', options)
      const y = result.y
      expect(y).toBe(result.y)
    })

    it('should return objects with useless setters', async () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = await evaluateAsync('x.y', options)
      result.z = 'qux'
      expect(await evaluateAsync('x.y.z', options)).toBe('foo')
    })

    it('should ignore rules on non-object, non-function scopes', async () => {
      const options = {
        explicitAllow: true,
        rules: [],
        scope: 'str' as any,
      }
      expect(await evaluateAsync('length', options)).toBe(3)
    })
  })

  describe('with explicit allow set to false', () => {
    it('should be able to access properties w/o matching rule', async () => {
      const options = {
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('foo', options)).toEqual({
        bar: { baz: 'quux' },
      })
      expect(await evaluateAsync('foo.bar', options)).toEqual({ baz: 'quux' })
      expect(await evaluateAsync('foo["b" + "a" + "r"]', options)).toEqual({
        baz: 'quux',
      })
      expect(await evaluateAsync('alpha[symbolBeta]', options)).toBe('baz')
    })

    it('should be able to access properties w/ matching allow rule', async () => {
      const options = {
        rules: [
          { allow: 'x.y.z' },
          { block: 'a.b' },
          { allow: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('x.y.z', options)).toBe('foo')
      expect(await evaluateAsync('x["y"]["z"]', options)).toBe('foo')
      expect(await evaluateAsync('alpha[symbolBeta]', options)).toBe('baz')
    })

    it('should not be able to access properties w/ matching block rule', async () => {
      const options = {
        rules: [
          { allow: 'a' },
          { block: 'a.b' },
          { block: ['alpha', symbolBeta] },
        ],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('a.b', options)).toBeUndefined()
      expect(await evaluateAsync('a["b"]', options)).toBeUndefined()
      expect(async () => evaluateAsync('a.b.c', options)).rejects.toThrowError()
      expect(async () =>
        evaluateAsync('a["b"]["c"]', options)
      ).rejects.toThrowError()
      expect(await evaluateAsync('alpha[symbolBeta]', options)).toBeUndefined()
    })

    it('should return objects that return the same instances on subsequent property access', async () => {
      const options = {
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = await evaluateAsync('x', options)
      const y = result.y
      expect(y).toBe(result.y)
    })

    it('should return objects with useless setters', async () => {
      const options = {
        rules: [{ allow: 'x.y.z' }, { block: 'a.b' }],
        scope: getTestObject(),
      }
      const result = await evaluateAsync('x.y', options)
      result.z = 'qux'
      expect(await evaluateAsync('x.y.z', options)).toBe('foo')
    })

    it('should ignore rules on non-object, non-function scopes', async () => {
      const options = {
        rules: [{ block: 'length' }],
        scope: 'str' as any,
      }
      expect(await evaluateAsync('length', options)).toBe(3)
    })
  })

  describe('Rule parsing', () => {
    it('should support escaped wildcards', async () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'wilds.\\*' }, { allow: 'wilds.\\**' }],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('wilds["*"]', options)).toBe('quuz')
      expect(await evaluateAsync('wilds["**"]', options)).toBe('corge')
    })

    it('should ignore falsy rules', async () => {
      const options = {
        rules: [undefined as any],
        scope: getTestObject(),
      }
      expect(async () => evaluateAsync('x.y.z', options)).not.toThrow()
    })

    it('should ignore rules with empty paths', async () => {
      const options = {
        rules: [{ block: '' }, { block: [] }],
        scope: getTestObject(),
      }
      expect(async () => evaluateAsync('x.y.z', options)).not.toThrow()
    })

    it('should respect rule order', async () => {
      const options = {
        explicitAllow: true,
        rules: [{ allow: 'a.b.c' }, { block: 'a.b' }, { allow: 'a.b' }],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('a.b.c', options)).toBeUndefined()
      expect(await evaluateAsync('a.b', options)).toEqual({ c: undefined })
    })

    it('should support wildcards', async () => {
      const options = {
        explicitAllow: true,
        rules: [
          { allow: 'a.**' },
          { block: 'a.*.c' },
          { allow: 'shallow.**' },
          { block: 'shallow.**.grault' },
          { allow: 'shallow.**.deepest.waldo' },
        ],
        scope: getTestObject(),
      }
      expect(await evaluateAsync('a.b.c', options)).toBeUndefined()
      expect(await evaluateAsync('a.b', options)).toEqual({ c: undefined })
      expect(
        await evaluateAsync('shallow.deep.deeper.deepest.grault', options)
      ).toBeUndefined()
      expect(
        await evaluateAsync('shallow.deep.deeper.deepest.waldo', options)
      ).toBe('fred')
    })

    it("should throw an error if a rule's path is not a string or string array", () => {
      const options = {
        rules: [{ allow: true as any }],
        scope: getTestObject(),
      }
      expect(async () =>
        evaluateAsync('x.y.z', options)
      ).rejects.toThrowErrorMatchingSnapshot()
    })

    it("should throw an error if a rule's path part is not a string or a symbol", () => {
      const options = {
        rules: [{ allow: ['x', true as any] }],
        scope: getTestObject(),
      }
      expect(async () =>
        evaluateAsync('x.y.z', options)
      ).rejects.toThrowErrorMatchingSnapshot()
    })
  })
})
