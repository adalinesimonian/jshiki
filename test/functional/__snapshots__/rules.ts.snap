// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rules Rule parsing should throw an error if a rule's path is not a string or string array 1`] = `"Path must be a string or an array of strings and/or symbols, got boolean"`;

exports[`Rules Rule parsing should throw an error if a rule's path part is not a string or a symbol 1`] = `"Path part must be a string or a symbol, got boolean"`;

exports[`Rules with explicit allow set to false should not be able to access properties w/ matching block rule 1`] = `"Cannot read property 'c' of undefined"`;

exports[`Rules with explicit allow set to false should not be able to access properties w/ matching block rule 2`] = `"Cannot read property 'c' of undefined"`;

exports[`Rules with explicit allow set to true should not be able to access properties w/ matching block rule 1`] = `"Cannot read property 'c' of undefined"`;

exports[`Rules with explicit allow set to true should not be able to access properties w/ matching block rule 2`] = `"Cannot read property 'c' of undefined"`;

exports[`Rules with explicit allow set to true should not be able to access properties w/o matching rule 1`] = `"Cannot read property 'bar' of undefined"`;

exports[`Rules with explicit allow set to true should not be able to access properties w/o matching rule 2`] = `"Cannot read property 'bar' of undefined"`;

exports[`Rules with explicit allow set to true should not be able to access properties w/o matching rule 3`] = `"Cannot read property 'Symbol(beta)' of undefined"`;
